juju/juju
# [JUJU-4126] Add UpdateExternalControllerModel and replace the usages of SaveAndMoveModels on facades

This patch adds a new domain method for external controllers: UpdateExternalControllerModel, which updates the external_model table with the model UUID and the controller UUID it is linked to.

This is a first step needed to completely remove the legacy mongodb state for external controllers.

## Checklist


## QA steps

First, let's create two controllers (one with influxdb, other with grafana) and integrate them:
```
juju bootstrap localhost ctrl-src1
juju add-model model-src1
juju deploy influxdb
juju offer influxdb:grafana-source  
---
juju bootstrap localhost ctrl-src2
juju add-model model-src2
juju deploy grafana
juju integrate grafana ctrl-src1:admin/model-src1.influxdb
```
Now you can check the status with `--relations` to validate that the integration worked.

Also, let's check that the external controller have been stored in dqlite and not in mongodb:
(still on ctrl-src2)
connect to controller machine and run dqlite:
```
lxc exec juju-xxx bash
apt install go-dqlite
dqlite -s 127.0.0.1:17666 controller
SELECT * FROM external_controller; // Here you should see one external controller:
dqlite> SELECT * FROM external_controller;
b350dd86-cff7-4e42-8979-f8b40d5907e0|ctrl-src1|-----BEGIN CERTIFICATE-----
MIIEEzCCAnugAwIBAgIVALivOnMNMuM2SgZNOFE1dgDwKoZ3MA0GCSqGSIb3DQEB
CwUAMCExDTALBgNVBAoTBEp1anUxEDAOBgNVBAMTB2p1anUtY2EwHhcNMjMwNzI0
MDcyNzU3WhcNMzMwNzI0MDczMjU3WjAhMQ0wCwYDVQQKEwRKdWp1MRAwDgYDVQQD
EwdqdWp1LWNhMIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEA0/uqd0Ac
aINPKp/lhdCKDAfBFKK5HKIt0GmSXBWa0xHwOwKi7a/h3qUhwAG4xJnLGGZc0WUY
mT8ciaJhM9uH41oopkLOvZnI8GU+D+srb1uyvYHIp9v/yICy/K61glUyoiu4ckf/
KeiPPXEZbd3u2Jz9roZPTvPgfes9Otuft3+82Dx6oW8vA4WZb/CGnRZ/tGr9oQ4M
Qqe+TbNM/kxeafyQ2c0yjVG2v3fdaAVT7PE0d8uiWQxVvcQGAyPHr2RPhhqxILJ+
YXTwKo+v46TlJj33GBKUQJfA2oGbtyv9Y7ybkhWfFsWIEmmckZU0Km97N9aGtnqX
NPF1Q85WG3OmO+odiYQ/nDpC/kL9ohndEA1HK7rzOLIk/VynHXEgzMK1Oi5U9aJH
rZ6eecyqKJ5QOkvsBCiA3zzsqRnIr/xAxnsoVTHmlIHaWU/b0hA/l+9jin+1OpbC
H3e14faaDewcsejBP5VdQjTpM+fo8Hni1lnZutxBB2PXt6yNlYt7HqaPAgMBAAGj
QjBAMA4GA1UdDwEB/wQEAwICpDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBTJ
xfk7RLSg65nniiCdl67BsyYu5DANBgkqhkiG9w0BAQsFAAOCAYEAEJAl0cc+K45q
KFF8bExhWW9NJ4gLsmgBNmosnlVHv4UFWEElpvIfN8Ieoc0dF5V9jy7uWMx/SCI3
Q/ZHW/xJdic0dGXkzBTze3INT43DLrI/3bFjPlfHiKg9+vcdLhLLbF8Hv9hgAubS
UB+qaAl8QUUZ0MbXDXU+C4gfUS9aMb1HFVbgkf3FAQkftDERziKzexkWpsLyJmHS
Zviqje2YA9gYvAQ/t40mYbomvzwdXi11WVO3VdfvRv/iR6YutXO04wZZIJNDYyHM
+AJ+q3fuonRFgk06e4VZybLyGan4KHIcYyr9eJ71Bf72m71lz2yiiZ+AcnI0Nx9v
7PCR5lrQArF4XXBauUnAkC16FGFdAByxHyVKt5/1Es39aZmIbrNcyrHhPznl/ozt
5Z/xwJ+hMWZtYoCB0qtnll3ua1/on4Ho9l6AesJ/7b+BQRarcbdkPJ/sr1UAL39V
m6G3JGIcyHJnols6PoXQMB2Q92hY39b0XhXhQ5DiL/9XfueM96l1
-----END CERTIFICATE-----
```
And mongodb:
```
juju:PRIMARY> db.externalControllers.count()
0
```

Now let's bootstrap a new controller and migrate model-src2 (grafana integrated with influxdb) to it:
```
juju bootstrap localhost ctrl-target
juju switch ctrl-src2
juju migrate model-src2 ctrl-target
```
Let's check that everything has been migrated correctly and stored in dqlite and not in mogodb:
```
juju switch ctrl-target
juju switch model-src2

lxc exec juju-xxx bash

apt install go-dqlite
dqlite -s 127.0.0.1:17666 controller

dqlite> SELECT * FROM external_controller;
b350dd86-cff7-4e42-8979-f8b40d5907e0|ctrl-src1|-----BEGIN CERTIFICATE-----
MIIEEzCCAnugAwIBAgIVALivOnMNMuM2SgZNOFE1dgDwKoZ3MA0GCSqGSIb3DQEB
CwUAMCExDTALBgNVBAoTBEp1anUxEDAOBgNVBAMTB2p1anUtY2EwHhcNMjMwNzI0
MDcyNzU3WhcNMzMwNzI0MDczMjU3WjAhMQ0wCwYDVQQKEwRKdWp1MRAwDgYDVQQD
EwdqdWp1LWNhMIIBojANBgkqhkiG9w0BAQEFAAOCAY8AMIIBigKCAYEA0/uqd0Ac
aINPKp/lhdCKDAfBFKK5HKIt0GmSXBWa0xHwOwKi7a/h3qUhwAG4xJnLGGZc0WUY
mT8ciaJhM9uH41oopkLOvZnI8GU+D+srb1uyvYHIp9v/yICy/K61glUyoiu4ckf/
KeiPPXEZbd3u2Jz9roZPTvPgfes9Otuft3+82Dx6oW8vA4WZb/CGnRZ/tGr9oQ4M
Qqe+TbNM/kxeafyQ2c0yjVG2v3fdaAVT7PE0d8uiWQxVvcQGAyPHr2RPhhqxILJ+
YXTwKo+v46TlJj33GBKUQJfA2oGbtyv9Y7ybkhWfFsWIEmmckZU0Km97N9aGtnqX
NPF1Q85WG3OmO+odiYQ/nDpC/kL9ohndEA1HK7rzOLIk/VynHXEgzMK1Oi5U9aJH
rZ6eecyqKJ5QOkvsBCiA3zzsqRnIr/xAxnsoVTHmlIHaWU/b0hA/l+9jin+1OpbC
H3e14faaDewcsejBP5VdQjTpM+fo8Hni1lnZutxBB2PXt6yNlYt7HqaPAgMBAAGj
QjBAMA4GA1UdDwEB/wQEAwICpDAPBgNVHRMBAf8EBTADAQH/MB0GA1UdDgQWBBTJ
xfk7RLSg65nniiCdl67BsyYu5DANBgkqhkiG9w0BAQsFAAOCAYEAEJAl0cc+K45q
KFF8bExhWW9NJ4gLsmgBNmosnlVHv4UFWEElpvIfN8Ieoc0dF5V9jy7uWMx/SCI3
Q/ZHW/xJdic0dGXkzBTze3INT43DLrI/3bFjPlfHiKg9+vcdLhLLbF8Hv9hgAubS
UB+qaAl8QUUZ0MbXDXU+C4gfUS9aMb1HFVbgkf3FAQkftDERziKzexkWpsLyJmHS
Zviqje2YA9gYvAQ/t40mYbomvzwdXi11WVO3VdfvRv/iR6YutXO04wZZIJNDYyHM
+AJ+q3fuonRFgk06e4VZybLyGan4KHIcYyr9eJ71Bf72m71lz2yiiZ+AcnI0Nx9v
7PCR5lrQArF4XXBauUnAkC16FGFdAByxHyVKt5/1Es39aZmIbrNcyrHhPznl/ozt
5Z/xwJ+hMWZtYoCB0qtnll3ua1/on4Ho9l6AesJ/7b+BQRarcbdkPJ/sr1UAL39V
m6G3JGIcyHJnols6PoXQMB2Q92hY39b0XhXhQ5DiL/9XfueM96l1
-----END CERTIFICATE-----
```
and mongodb:
```
juju:PRIMARY> db.externalControllers.count()
0
```
